Improve Molten Core Strategy

HennyWilly opened the pull request:
This is my first attempt of implementing playerbot strategies.

A team of 40 can steamroll Molten Core relatively easy, even with lower item levels.
Regardless, this PR adds resistance triggers and actions to mitigate some damage.
Additionally, improvements were made for the encounters with Garr, Baron Geddon, Shazzrah and Golemagg.
A short summary per boss is listed below.
All planned features are included, but feedback is of course appreciated.

### Lucifron
- Shadow resistance: mitigate damage from [Impending Doom](https://www.wowhead.com/classic/spell=19702/impending-doom) and [Shadow Shock](https://www.wowhead.com/classic/spell=19460/shadow-shock).

### Magmadar
- Fire resistance: mitigate damage from [Lava Bomb](https://www.wowhead.com/classic/spell=19411/lava-bomb) and [Magma Spit](https://www.wowhead.com/classic/spell=19450/magma-spit).
- Like King Dred and the fraction commander (Nexus), this fight might profit from an anti-fear strategy in order to counter [Panic](https://www.wowhead.com/classic/spell=19408/panic). Not implemented here.

### Gehennas
- Shadow resistance: mitigate damage from [Shadow Bolt](https://www.wowhead.com/classic/spell=19728/shadow-bolt) and increase the chance to resist [Gehennas' Curse](https://www.wowhead.com/classic/spell=19716/gehennas-curse).

### Garr
- Fire resistance: mitigate damage from the Firesworn adds ([Immolate](https://www.wowhead.com/classic/spell=20294/immolate) and [Eruption](https://www.wowhead.com/classic/spell=19497/eruption)).
-  Disabled dps aoe abilities via multiplier. This one is important because multiple exploding adds at once might delete bots rather quick...

### Baron Geddon
- Refactored the existing strategy.
- Fire resistance: mitigate damage from [Ignite Mana](https://www.wowhead.com/classic/spell=19659/ignite-mana), [Inferno](https://www.wowhead.com/classic/spell=19695/inferno) and [Living Bomb](https://www.wowhead.com/classic/spell=20475/living-bomb).
- Better Inferno handling: Before moving away, bots stop attacking and interrupt their spells. Additionally, the new multiplier prevents bots from running back to Geddon while Inferno is still active.

### Shazzrah
- Ranged bots now position themselves in a sweet spot that prevents them from getting hit with [Arcane Explosion](https://www.wowhead.com/classic/spell=19712/arcane-explosion) but still close enough to dps and heal.

### Sulfuron Harbinger
- Fire resistance: mitigate damage from [Hand of Ragnaros](https://www.wowhead.com/classic/spell=19780/hand-of-ragnaros) and [Immolate](https://www.wowhead.com/classic/spell=20294/immolate). To be fair, this one is quite negligible...

### Golemagg
- Fire resistance: mitigate damage from [Magma Splash](https://www.wowhead.com/classic/spell=13880/magma-splash) and [Pyroblast](https://www.wowhead.com/classic/spell=20228/pyroblast).
-  Disabled dps aoe abilities via multiplier. Kind of a preference on my side. Otherwise, the Core Ragers spam emotes about not wanting to die.

### Majordomo Executus
- Shadow resistance: mitigate damage from [Aegis of Ragnaros](https://www.wowhead.com/classic/spell=20620/aegis-of-ragnaros), [Shadow Shock](https://www.wowhead.com/classic/spell=20603/shadow-shock) and [Shadow Bolt](https://www.wowhead.com/classic/spell=21077/shadow-bolt). This one is also negligible, TBF.

### Ragnaros
- Fire resistance: mitigate damage from [Wrath of Ragnaros](https://www.wowhead.com/classic/spell=20566/wrath-of-ragnaros) and [Lava Burst](https://www.wowhead.com/classic/spell=21158/lava-burst).

Celandriel commented on src/strategy/raids/moltencore/RaidMcMultipliers.h:
```diff
@@ -0,0 +1,27 @@
+#ifndef _PLAYERBOT_RAIDGRUULSLAIRMULTIPLIERS_H
```
Change to _PLAYERBOT_RAIDMCMULTIPLIERS_H

Celandriel commented on src/strategy/raids/moltencore/RaidMcActions.h:
```diff
@@ -1,6 +1,7 @@
 #ifndef _PLAYERBOT_RAIDMCACTIONS_H
 #define _PLAYERBOT_RAIDMCACTIONS_H
 
+#include "AttackAction.h"
```
Is this being used?

Celandriel replied:
Is this needed here? I dont see any Attack Actions referenced.

Celandriel commented on src/strategy/raids/moltencore/RaidMcMultipliers.h:
```diff
@@ -0,0 +1,27 @@
+#ifndef _PLAYERBOT_RAIDGRUULSLAIRMULTIPLIERS_H
+#define _PLAYERBOT_RAIDGRUULSLAIRMULTIPLIERS_H
```
_PLAYERBOT_RAIDMCMULTIPLIERS_H

HennyWilly replied:
Changed to _PLAYERBOT_RAIDMCMULTIPLIERS_H

Celandriel commented on src/strategy/raids/moltencore/RaidMcHelpers.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef RAID_GRUULSLAIRHELPERS_H
+#define RAID_GRUULSLAIRHELPERS_H
```
RAID_MCHELPERS_H

HennyWilly replied:
Changed to _PLAYERBOT_RAIDMCHELPERS_H for consistent naming.

brighton-chi commented on src/strategy/raids/moltencore/RaidMcStrategy.h:
```diff
@@ -9,9 +9,9 @@ class RaidMcStrategy : public Strategy
-    virtual std::string const getName() override { return "mc"; }
-    virtual void InitTriggers(std::vector<TriggerNode*>& triggers) override;
-    // virtual void InitMultipliers(std::vector<Multiplier*> &multipliers) override;
+    std::string const getName() override { return "mc"; }
```
This is just my opinion, but I think we should take the opportunity to rename the Molten Core strategy from "mc" to "moltencore" (which would require corresponding changes in PlayerbotAI.cpp and RaidStrategyContext.h). The name "mc" is just too vague imo--I think it's better for co ? or nc ? to show "moltencore" instead because it is much clearer, and such strategy name is not that long anyway.

HennyWilly replied:
Makes sense. Renamed "mc" to "moltencore" in this file, PlayerbotAI.cpp and RaidStrategyContext.h.

brighton-chi commented:
Thanks for doing this--love to see more work on boss strategies.

I wanted to bring this up (for you and approvers), but I've always thought the MC strategy is suboptimal because it toggles bot strategies. That's something I think generally to be avoided because it can persist, and I have had the issue with Baron Geddon where bots that have Living Bomb when Geddon dies will stay in place after he dies. I don't know exactly what causes it, but my suspicion is the usage of a bot strategy is the issue.

Ultimately, I don't understand why the move from group strategy needs to exist. I would have thought it would be fine to directly call MoveFromGroup (I've never used it though so not sure what it does), or use MoveAway or a custom MoveTo that tries to account for all group members at once. Then you simply have an action that executes at the applicable time in combat and has no effect outside of the active encounter.

HennyWilly commented on src/strategy/raids/moltencore/RaidMcActionContext.h:
```diff
@@ -10,13 +10,33 @@ class RaidMcActionContext : public NamedObjectContext<Action>
 public:
     RaidMcActionContext()
     {
+        creators["mc lucifron shadow resistance action"] = &RaidMcActionContext::lucifron_shadow_resistance_action;
```
Looks like most dungeon/raid strategies don't use the "action" and "trigger" suffix. Removed.

HennyWilly commented on src/strategy/raids/moltencore/RaidMcTriggers.cpp:
```diff
@@ -1,22 +1,30 @@
     // only do when fighting baron geddon (to avoid modifying strat set by player outside this fight)
     if (Unit* boss = AI_VALUE2(Unit*, "find target", "baron geddon"))
     {
         if (boss->IsInCombat())
```
Check removed.

HennyWilly commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,12 +1,21 @@
+    if (bot->HasAura(SPELL_LIVING_BOMB)) // baron geddon's living bomb
     {
-        if (!botAI->HasStrategy("move from group", BotState::BOT_STATE_COMBAT))
+        if (!botAI->HasStrategy("move from group", BOT_STATE_COMBAT))
```
[See my comment below.](https://github.com/mod-playerbots/mod-playerbots/pull/1852#issuecomment-3549728491)

brighton-chi commented:
You could try just having the bots mark Golemagg with a skull. I am just guessing here, but the issue might be that if you disable AoE actions, a bunch of bots end up single targeting dogs instead.


I think marking outside of contexts where targets need to be split should generally be a limited strategy because it takes away player agency, but for a situation like this where there is no sane reason to target the dogs, it's probably a good idea.

isitsimon commented:
> You could try just having the bots mark Golemagg with a skull. I am just guessing here, but the issue might be that if you disable AoE actions, a bunch of bots end up single targeting dogs instead.
> 
> I think marking outside of contexts where targets need to be split should generally be a limited strategy because it takes away player agency, but for a situation like this where there is no sane reason to target the dogs, it's probably a good idea.

While I agree, the dogs need to be tanked. There is also a concern that 2 buffed dogs with 100% increased damage + the boss all agro to one tank. 

That tank will not last long 🤣. 

Ideally an off tank would take the dogs away and MT would sit on boss.

brighton-chi commented:
> > You could try just having the bots mark Golemagg with a skull. I am just guessing here, but the issue might be that if you disable AoE actions, a bunch of bots end up single targeting dogs instead.
> 
> > 
> 
> > I think marking outside of contexts where targets need to be split should generally be a limited strategy because it takes away player agency, but for a situation like this where there is no sane reason to target the dogs, it's probably a good idea.
> 
> 
> 
> While I agree, the dogs need to be tanked. There is also a concern that 2 buffed dogs with 100% increased damage + the boss all agro to one tank. 
> 
> 
> 
> That tank will not last long 🤣. 
> 
> 
> 
> Ideally an off tank would take the dogs away and MT would sit on boss. 
> 
> 

I see. The simplest solution without needing to split targets might be to have a different mark on Golemagg then and then have DPS prioritize that mark but don't make changes with respect to tank RTI.

Then the player can still control the tanks separately, but DPS will attack only Golemagg.

Celandriel commented:
@HennyWilly Checking if you are still working on this, or if I should review.

noisiver commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,43 +1,220 @@
+
+        float orientation = atan2(target->GetPositionY() - bot->GetPositionY(),
+                                  target->GetPositionX() - bot->GetPositionX());
+        bot->SetFacingTo(orientation);
```
Is it really necessary to force their orientation? Wouldn't it make more sense to let them face the boss on their own since it doesn't actually affect the boss in any way?

noisiver replied:
Yes but this is setting the **bot** orientation, not the boss orientation. That's why it confuses me. The boss will always face the player (or bot) so does the bots orientation ever actually matter?

brighton-chi replied:
In my experience it did, but it's not like I tested this specific logic many times so it could just be up to chance. It is possible that the orientation could influence how the bot chooses to engage the boss, but I can't say anything definitive.

noisiver commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,43 +1,220 @@
+                                  target->GetPositionX() - bot->GetPositionX());
+        bot->SetFacingTo(orientation);
+    }
+    else if (!bot->IsWithinMeleeRange(target))
```
Don't they do this on their own? They always run up the target they're attacking, is there a reason why you're manually handling this on top of that?

brighton-chi commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,43 +1,220 @@
+    if (Unit* boss = AI_VALUE2(Unit*, "find target", "baron geddon"))
     {
-        if (!botAI->HasStrategy("move from group", BotState::BOT_STATE_COMBAT))
+        float distToTravel = INFERNO_DISTANCE - bot->GetDistance(boss);
```
Do you know if the AoE range is measured in three dimensions? I prefer GetDistance2d (if accounting for the target's size) or GetExactDist2d (if measuring from the target's middle). It's not clear to me that AC handles AoE radii correctly as it seems to measure at least some spells from the edge of the caster's hitbox (but I can't get a clear answer on what is right vs. wrong). I don't know about the 2d vs. 3d matter, though.

brighton-chi commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,43 +1,220 @@
+            float dist = sqrt(dX * dX + dY * dY);
+            float moveX = bot->GetPositionX() + (dX / dist) * maxDistance;
+            float moveY = bot->GetPositionY() + (dY / dist) * maxDistance;
+            return MoveTo(bot->GetMapId(), moveX, moveY, bot->GetPositionZ(), false, false,
```
You can use a constant for the Molten Core map ID and replace all GetMapId() with the constant to avoid the lookup.

HennyWilly commented:
Here is an updated version. The comments above about `SetFacingTo` and `IsWithinMeleeRange` have been addressed.

I also fixed an issue that prevented assist tank 2 from picking up the free Core Rager if assist tank 0 or 1 died. The problem was that `IsAssistTankOfIndex` doesn't take dead players/bots into account.
For example, if assist tank 1 dies, assist tank 2 would not be "promoted" and `IsAssistTankOfIndex(bot, 1)` would still return false.
To address this, I added an optional parameter `bool ignoreDeadPlayers = false` to `IsAssistTankOfIndex`.
For the above example, with this parameter set to true, `IsAssistTankOfIndex(bot, 1, true)` would return true for assist tank 2 if either assist tank 0 or 1 is dead.
This should not break existing strategies, but I understand that this is a change outside of the Molten Core strategy. I will think of something else if this change is not desired.

Lastly, the bots didn't seem to use their taunt abilities, so I added `botAI->DoSpecificAction("taunt spell", ...)`. With this addition, the bots will use their class specific taunt abilities if their respective targets don't attack them back.
Death Knights learn their taunt ability [Dark Command](https://www.wowhead.com/wotlk/spell=56222) at level 65, so DK tanks below that level may require a little more time to (re-)gain aggro.
If there is a better way to do this, let me know.

brighton-chi commented:
Yeah I believe @nl-saw is going to modify the various assistant roles to make them exclude dead bots but for now it makes sense to manually.

I'll try to take a look at the overall strategy.

brighton-chi commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,43 +1,218 @@
+    if (Unit* boss = AI_VALUE2(Unit*, "find target", "baron geddon"))
     {
-        if (!botAI->HasStrategy("move from group", BotState::BOT_STATE_COMBAT))
+        float distToTravel = INFERNO_DISTANCE - bot->GetDistance(boss);
```
Unless the radius for this attack should be measured in 3d, I would use a 2d distance measurement (either GetDistance2d if taking the boss's hitbox into account or GetExactDist2d if not).

brighton-chi commented on src/strategy/raids/moltencore/RaidMcActions.h:
```diff
@@ -21,4 +22,46 @@ class McMoveFromBaronGeddonAction : public MovementAction
+    McGolemaggTankAction(PlayerbotAI* botAI, std::string const name)
+        : AttackAction(botAI, name) {}
+protected:
+    bool MoveUnitToPosition(Unit* target, const Position& tankPosition, float maxDistance, float stepDistance = 3.f);
```
Typo for stepDistance?

brighton-chi commented on src/strategy/raids/moltencore/RaidMcStrategy.cpp:
```diff
@@ -1,13 +1,81 @@
+                        NextAction::array(0, new NextAction("mc lucifron shadow resistance", ACTION_RAID), nullptr)));
+
+    // Magmadar
+    // TODO: Fear ward / tremor totem, or general anti-fear strat development. Same as King Dred (Drak'Tharon) and faction commander (Nexus).
```
Are you planning this still? FWIW, since tremor totem is a class strategy now, I don't think you really need a raid strategy to use it.

brighton-chi commented on src/strategy/raids/moltencore/RaidMcMultipliers.cpp:
```diff
@@ -0,0 +1,115 @@
+    return true;
+}
+
+float BaronGeddonAbilityMultiplier::GetValue(Action* action)
```
My bad, I thought the multiplier was referencing a method, but I see now it's just a helper.

HennyWilly commented on src/strategy/raids/moltencore/RaidMcActions.cpp:
```diff
@@ -1,43 +1,218 @@
+        {
+            float dX = tankPosition.GetPositionX() - bot->GetPositionX();
+            float dY = tankPosition.GetPositionY() - bot->GetPositionY();
+            float dist = sqrt(dX * dX + dY * dY);
```
Yes, it looks like this code does exactly the same as the above `GetExactDist2d`.
And yes, the Gruul strategies do the same thing. 

Would you mind keeping this as is for now? 
I already have two approvals, and pushing further changes might reset them.  
The only remaining blocker is the outstanding "requested change", which is already resolved on my side.

isitsimon commented:
This is awesome HennyWilly. 👌🏻👏🏻

noisiver commented on src/PlayerbotAI.cpp:
```diff
@@ -2266,6 +2265,11 @@ bool PlayerbotAI::IsAssistTankOfIndex(Player* player, int index)
             continue;
         }
 
+        if (ignoreDeadPlayers && !member->IsAlive())
```
We aim to match the code style of azerothcore and the enforcement of it is still in progress so it should be single line if = no curly brackets. Because of that any code you add should follow that, no matter what the rest of the code looks like.

noisiver commented on src/PlayerbotAI.cpp:
```diff
@@ -2285,6 +2289,11 @@ bool PlayerbotAI::IsAssistTankOfIndex(Player* player, int index)
         }
 
+        if (ignoreDeadPlayers && !member->IsAlive())
+        {
```
Don't use curly brackets for single line if
